File: tests/zenko_tests/e2e_config/accounts.py
#!/usr/bin/env python

import logging

from kubernetes import client, config
from kubernetes.client.rest import ApiException

from e2e_config import clients

_log = logging.getLogger("end2end configuration")

def get_credentials(token, account_id):
    """
    Retrieve credentials from sts service

    :param token: oidc id token
    :param account_id: account id
    """
    _log.info("getting account credentials")

    res = clients.stsclient.assume_role_with_web_identity(
        RoleArn="arn:aws:iam::%s:role/roleForB" % (account_id),
        RoleSessionName='end2end',
        WebIdentityToken=token,
        DurationSeconds=60 * 60, # 1 hr
    )

    return res


def create_account_secret(name, credentials, namespace="default"):
    """
    Create a k8s secret resource for account

    :param name: secret name
    :param credentials: sts assume role credentials
    :param namespace: k8s namespace
    """
    _log.info("creating account secret")

    core = client.CoreV1Api()
    secret = client.V1Secret(
        api_version="v1",
        metadata=client.V1ObjectMeta(
            name=name,
            labels={
                "type": "end2end",
            },
        ),
        string_data=credentials,
    )

    res = core.create_namespaced_secret(namespace, body=secret)

    _log.info("created account secret")


def create_account(client, token, uuid, account_name, namespace="default"):
    """
    Creates a user account and save its accessKey and secretKey as k8s secret

    :param client: swagger client
    :param uuid: zenko instance uuid
    :param account_name: account_name
    :param namespace: k8s namespace
    """
    try:
        User_V1 = client.get_model('user-v1')
        u = User_V1(userName=account_name,
                    email="%s@zenko.local" % (account_name))

        res = (
            client.ui_facing
            .createConfigurationOverlayUser(user=u, uuid=uuid)
            .response()
            .result
        )

        creds = get_credentials(token, res.id)
        create_account_secret(name="end2end-account-%s" % (res.userName),
                              credentials=creds["Credentials"],
                              namespace=namespace)

        _log.info("created account")
    except Exception as e:
        raise Exception(
            "Failed to create account '%s': %s" % (account_name, e))

File: tests/functional/testResourcesSetup/setup/accounts.ts
import { STSClient, AssumeRoleWithWebIdentityCommand } from '@aws-sdk/client-sts';
import { CoreV1Api } from '@kubernetes/client-node';
import { PensieveClient } from './clients/pensieveClient';
import { AccountCredentials, createKubernetesSecret } from './clients/k8s';
import { Env } from '../config';

async function getAccountCredentials(
    stsClient: STSClient,
    token: string,
    accountId: string,
): Promise<AccountCredentials> {
    const result = await stsClient.send(new AssumeRoleWithWebIdentityCommand({
        RoleArn: `arn:aws:iam::${accountId}:role/scality-internal/storage-manager-role`,
        RoleSessionName: 'end2end',
        WebIdentityToken: token,
        DurationSeconds: 60 * 60 * 12, // 12 hours, max allows by sts assume role
    }));
    if (!result.Credentials?.AccessKeyId || !result.Credentials?.SecretAccessKey || !result.Credentials?.SessionToken) {
        throw new Error(`Incomplete credentials returned from STS for account ${accountId}`);
    }
    return {
        AccessKeyId: result.Credentials.AccessKeyId,
        SecretAccessKey: result.Credentials.SecretAccessKey,
        SessionToken: result.Credentials.SessionToken,
    };
}

export async function createAccounts(
    pensieveClient: PensieveClient,
    stsClient: STSClient,
    coreClient: CoreV1Api,
    env: Env,
    accountNames: string[],
): Promise<Record<string, AccountCredentials>> {
    const creds: Record<string, AccountCredentials> = {};
    for (const accountName of accountNames) {
        const user = await pensieveClient.createUser(env.UUID, accountName);
        const credentials = await getAccountCredentials(stsClient, env.TOKEN, user.id);
        await createKubernetesSecret(coreClient, env.NAMESPACE, `end2end-account-${user.userName}`, {
            AccessKeyId: credentials.AccessKeyId,
            SecretAccessKey: credentials.SecretAccessKey,
            SessionToken: credentials.SessionToken,
            AccountId: user.id,
        });
        console.log(`Created account: ${accountName}`);
        creds[accountName] = credentials;
    }
    return creds;
}
